src/org/javarosa/xpath/expr/XPathFuncExpr.java

```diff
@@ -1335,4 +1360,37 @@ public Object pivot (DataInstance model, EvaluationContext evalContext, List<Obj
         return eval(model, evalContext);
 
     }
+
+    enum HashEncodingMethod {
+        HEX("hex") {
+            @Override
+            String encode(byte[] bytes) {
+                StringBuilder stringBuffer = new StringBuilder();
+                for (byte aResult : bytes)
+                    stringBuffer.append(Integer.toString((aResult & 0xff) + 0x100, 16).substring(1));
+                return stringBuffer.toString();
+            }
+        },
+        BASE64("base64") {
+            @Override
+            String encode(byte[] bytes) {
+                return DatatypeConverter.printBase64Binary(bytes);
+            }
+        };
+
+        private final String value;
+
+        HashEncodingMethod(String value) {
+            this.value = value;
+        }
+
+        static HashEncodingMethod from(String value) {
+            for(HashEncodingMethod candidate : values())
+                if (candidate.value.equals(value))
+                    return candidate;
+            throw new XPathTypeMismatchException("Unsupported hash encoding algorithm \"" + value + "\". Supported values are: \"hex\", \"base64\"");
```

Comment:
Can you avoid reimplementing Enum.valueOf? Maybe like this?
```
try {
    return valueOf(value.toUpperCase());
} catch (IllegalArgumentException ex) {
    throw new XPathTypeMismatchException("Unsupported hash encoding algorithm \"" + value + "\". Supported values are: \"hex\", \"base64\"");
}
```

Reply:
The rationale was twofold:
- Hide the generation and error handling inside the type
- Avoid coupling between the enum name and the actual algorithm name